### DISCUSSION TITLE
Environment tree

### DISCUSSION THREAD

**Discussion by spoltorak** [Kong/insomnia#5629] [Ideas]
At the moment Insomnia supports only one base configuration with sub-configurations.

In my work, I have to deal with multiple environments and tenants in each env so I need more than one base configuration or configurations with more than one nesting level.

This can be illustrated as:

- prod env
- -  tenant A
- - tenant B
- staging env
- - tenant C
- - tenant D
- test env
- - tenant E

Or:
- root env
- - env level A1
- - - env level A12
- - env level B1

**Comment by platelminto** [comment]
I think you can use folders for this (`Ctrl+Shift+N`). You can then right-click on the folder and change its environment, and folders can be nested inside other folders.

I personally use a combination of both folders and the 'global' environment.

**Comment by ryancocuzzo** [comment]
This is very important for scalability. The root environment is toggle-able, but the nested environments aren't. 

It avoids duplication. Since nested environments aren't toggle-able we need to duplicate the endpoints.

For example, let's say you have 
- a global var `baseurl` that you want to toggle depending on your environment (testing vs production)
- a feature-specific var `user` that you want to toggle based on what you're testing (admin user vs standard user)

Right now, this would look like:

*Root env:*
```
baseUrl = ...
```

*Folders: (Given current limitations)*
```
/Food
  - {{baseUrl}}/user/{{adminUser}}/pizza
  - {{baseUrl}}/food/{{adminUser}}/burger
 /Drinks
  - {{baseUrl}}/user/{{adminUser}}/pizza
  - {{baseUrl}}/food/{{adminUser}}/burger 
 
/Food
  - {{baseUrl}}/user/{{standardUser}}/pizza
  - {{baseUrl}}/food/{{standardUser}}/burger
 /Drinks
  - {{baseUrl}}/user/{{standardUser}}/pizza
  - {{baseUrl}}/food/{{standardUser}}/burger 
```

Versus *the folders ideally:*
```
/Food - ENV: { "user": "adminUser" }
  - {{baseUrl}}/user/{{user}}/pizza
  - {{baseUrl}}/food/{{user}}/burger
 /Drinks - ENV: { "user": "standardUser"  }
  - {{baseUrl}}/user/{{user}}/pizza
  - {{baseUrl}}/food/{{user}}/burger 
```

**Reply by platelminto** [reply]
I don't understand - your "folders ideally" is exactly what you can achieve right now? Each folder can have its own environment, where you can define the appropriate value for `user`. Or am I misunderstanding your example?

I generally do agree that there could be a nicer implementation of nested environments, but you can get 99% of the way there using folders.

**Reply by ryancocuzzo** [reply]
Each folder can have its own environment, but only the root folder can have several environments and toggle between them. That capability is just as important in nested folders as it is in the root folder. 
Updated the original comment.
